tests/chez/chez028/ExpressionParser.idr

```diff
@@ -0,0 +1,79 @@
+module Main
+
+import Control.Monad.Identity
+import Control.Monad.Trans
+
+import Data.Maybe
+import Data.Fin
+import Data.Nat
+import Data.String.Parser
+import Data.String.BuildExpressionParser
+
+%default partial
+
+addDigit : Nat -> Nat -> Nat
+addDigit num d = 10*num + d
+
+fromDigits : List Nat -> Nat
+fromDigits xs = foldl addDigit 0 xs
+
+digit : Parser Nat
+digit = do x <- satisfy isDigit
+           case fromChar x of
+                Nothing => P $ \s => do pure $ Fail s.pos "not a digit"
+                Just y => P $ \s => Id (OK y s)
+  where fromChar : Char -> Maybe Nat
+        fromChar '0' = Just 0
+        fromChar '1' = Just 1
+        fromChar '2' = Just 2
+        fromChar '3' = Just 3
+        fromChar '4' = Just 4
+        fromChar '5' = Just 5
+        fromChar '6' = Just 6
+        fromChar '7' = Just 7
+        fromChar '8' = Just 8
+        fromChar '9' = Just 9
+        fromChar _ = Nothing
+
+natural : Parser Nat
+natural = do x <- some digit
+             pure (fromDigits x)
+
+skip : Parser a -> Parser ()
+skip = map (const ())
+
+space : Parser Char
+space = satisfy isSpace
+
```

Comment:
Some of the commonly useful utility functions in the test could be moved to Data.String.Parser (or maybe in a new separate module).

Reply:
Aye, was thinking that. Any preference on adding them to Data.String.Parser, or maybe adding an additional Data.String.Parser.Combinators or something?

Reply:
Let's add them to Parser for now.

Reply:
Updated. I left `digit` and `natural` here as they're a little involved, but can also move those if you like.

Reply:
I think these ought to be moved too as they are so useful others will inevitably
need them. You can make `digit` a bit simpler by using a list of pairs of char +
meaning and relying on:

```idris
Data.List.lookup : Eq a => a -> List (a, b) -> Maybe b
```

Reply:
Aye, updated there. Switched `digit` to be a `Parser (Fin 10)` like in Lightyear, makes a bit more sense than for it to be a `Parser Nat`.